File: pset6/caesar/caesar.py
from sys import argv
from cs50 import get_string
key = -1

if len(argv) is 2:
    key = int(argv[1])

if len(argv) is not 2 or key < 0:
    print("usage : python caesar.py <non negative key>")
    exit(1)
else:
    # prompt user for plaintext
    pt = []
    ptext = get_string("plaintext: ")
    for i in range(len(ptext)):
        pt.append(ptext[i])
    l = len(pt)
    ct = []
    # loop over every element of plaintext
    for i in range(l):
        # encipher only alphabetic chars
        if pt[i].isalpha():
            # preserve case of the char
            if pt[i].isupper():
                # convert ascii index to alpha index
                a = ord(pt[i])
                a -= 65
                temp = (a + key) % 26 + 65
                ct.append(chr(temp))
            elif pt[i].islower():
                b = ord(pt[i])
                b -= 97
                tmp = (b + key) % 26 + 97
                ct.append(chr(tmp))
        # for non alpha char
        else:
            ct.append(pt[i])
    ctext = ''.join(ct)
    print(f"ciphertext: {ctext}")

File: pset2/caesar/caesar.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <cs50.h>

int main(int argc, string argv[])
{
    int key;//Declare key in main so we can use it anywhere in our program
    //check no of cmd line arguments
    if (argc == 2)
    {
        key = atoi(argv[1]);
    }
    if (argc != 2 || key < 0)
    {
        printf("error: enter a valid key (non negative) after ./caesar and single cmd line argument\n");
        //return 1 to main if argc != 2 or key is negative
        return 1;
    }
    else
    {
        //prompt user for plaintext
        string pt = get_string("plaintext: ");
        int l = strlen(pt);
        //declare a string or a character array to store ciphertext
        char ct[l];
        //loop over every element of plaintext
        for (int i = 0; i < l; i++)
        {
            //we need to encipher only alphabatical characters otherwise leave them as it is
            if (isalpha(pt[i]))
            {
                //preserve case of the ciphertext character
                if (isupper(pt[i]))
                {
                    //convert ascii index to alphabatical index
                    pt[i] -= 65;
                    //equation to encipher
                    ct[i] = (pt[i] + key) % 26;
                    //convert alphabatical index back to ascii index
                    ct[i] += 65;
                }
                if (islower(pt[i]))
                {
                    //convert ascii index to alphabatical index
                    pt[i] -= 97;
                    //equation to encipher
                    ct[i] = (pt[i] + key) % 26;
                    //convert alphabatical index back to ascii index
                    ct[i] += 97;
                }
            }
            else
            {
                //for non alphabatical characters
                ct[i] = pt[i];
            }
        }
        //print ciphertext with a new line
        printf("ciphertext: %s\n", ct);
    }
}